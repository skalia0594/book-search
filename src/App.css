.App {
  text-align: center;
  height: 100vh;
  display: flex;
  flex-direction: column;
}

header,
footer {
  max-height: 5rem;
  background-color: black;
  color: white;
}

main {
  flex: 1;
}
label {
  font-weight: bold;
}

input[type=text] {
  width: 50%;
  box-sizing: border-box;
  border: 2px solid #ccc;
  border-radius: 4px;
  font-size: 16px;
  background-color: white;
  background-position: 10px 10px; 
  background-repeat: no-repeat;
  text-align: center;
}

.button{
  background-color: #FFFFFF;
  border: 1px solid #222222;
  border-radius: 8px;
  box-sizing: border-box;
  color: #222222;
  cursor: pointer;
  display: inline-block;
  font-size: 14px;
  font-weight: 600;
  line-height: 12px;
  margin: 5px;
  outline: none;
  padding: 13px 23px;
  position: relative;
  text-align: center;
  text-decoration: none;
  user-select: none;
  -webkit-user-select: none;
  width: auto;
}

.button:focus-visible {
  box-shadow: #222222 0 0 0 2px, rgba(255, 255, 255, 0.8) 0 0 0 4px;
  transition: box-shadow .2s;
}

.button:active {
  background-color: #F7F7F7;
  border-color: #000000;
  transform: scale(.96);
}

.button:disabled {
  border-color: #DDDDDD;
  color: #DDDDDD;
  cursor: not-allowed;
  opacity: 1;
}

ul {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  list-style:none;
}

.media {
  display: flex;
  align-items: flex-start;
}
.media .content {
  flex: 1;
  padding: 10px;
}

.image {
  min-height: 284px;
  min-width: 180px;
}

.fields {
  display: flex;
  justify-content: space-between;
  max-width: fit-content;
  padding-bottom: inherit;
}
@media screen and (max-width: 480px) {
  .fields {
    display: flex;
    flex-direction: column;
  }
}