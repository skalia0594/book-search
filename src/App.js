import React, { useState } from 'react'
import './App.css'
import Search from './components/Search'
import List from './components/List'

const App = () => {
  // states
  const [searchKeyWord, setSearchKeyWord] = useState('')
  const [matchedRecord, setMatchedRecord] = useState([])
  const [isloading, setIsLoading] = useState(false)
  const [noRecordFound, setNoRecordFound] = useState(false)
  const [prevSortData, setPrevSortData] = useState([])
  const [sortByTitleFlag, setSortByTitleFlag] = useState(false)
  const [sortByYearFlag, setSortByYearFlag] = useState(false)
  const [disableTitleSort, setDisableTitleSort] = useState(false)
  const [disableYearSort, setDisableYearSort] = useState(false)
  // handlers
  const handleOnChangeInput = (e) => {
    setSearchKeyWord(e.target.value)
    setNoRecordFound(false)
  }
  const setIntialStates = () => {
    setSearchKeyWord('')
    setMatchedRecord([])
    setPrevSortData([])
    setSortByTitleFlag(false)
    setSortByYearFlag(false)
    setDisableTitleSort(false)
    setDisableYearSort(false)
  }

  const handleOnSubmit = (e) => {
    setIsLoading(true)
    fetchKeyWordData(searchKeyWord)
    .then(data => {
      setMatchedRecord(data)
      setIsLoading(false)
      return data
    })
    .then(data => data.length < 1 ? setNoRecordFound(true) : setNoRecordFound(false))
    .then(setIntialStates())
    .then(e.preventDefault())
  }
  // fetch data
  const fetchKeyWordData = async (key) => {
    if(key === '') {
      return []
    } 
    const apiUrl = `http://openlibrary.org/search.json?q=${key}&limit=10`
    const res = await fetch(apiUrl)
    const result = await res.json()
    const docs = await result.docs
    return docs
  }
  // sort fns
  const compareFn = (a, b) => {
    if (a < b) {
      return -1
    }
    if (a > b) {
      return 1
    }
    return 0
  }
  const sortTitle = () => {
    if(!sortByTitleFlag) {
      setPrevSortData([...matchedRecord])
      const sortByTitle = matchedRecord.sort((a,b) => {
        const titleA = a.title.toUpperCase()
        const titleB = b.title.toUpperCase()
        return compareFn(titleA, titleB)
      })
      setMatchedRecord(sortByTitle)
      setSortByTitleFlag(true)
      setDisableYearSort(true)
    }else {
      setMatchedRecord(prevSortData)
      setSortByTitleFlag(false)
      setDisableYearSort(false)
    }
  }
  const sortPublishYear = () => {
    if(!sortByYearFlag) {
      setPrevSortData([...matchedRecord])
      const sortByPublishYear = matchedRecord.sort((a, b) => {
        const yearA = a.first_publish_year
        const yearB = b.first_publish_year
        return compareFn(yearB, yearA)
      })
      setMatchedRecord(sortByPublishYear)
      setSortByYearFlag(true)
      setDisableTitleSort(true)
    }else{
      setMatchedRecord(prevSortData)
      setSortByYearFlag(false)
      setDisableTitleSort(false)
    }
    
  }

  return (
    <div className="App">
      <header>
        <Search 
          searchKeyWord={searchKeyWord}
          handleOnChangeInput={handleOnChangeInput}
          handleOnSubmit={handleOnSubmit}
        />
      </header>
      
      <main>
        <List 
          isloading={isloading}
          matchedRecord={matchedRecord}
          disableTitleSort={disableTitleSort}
          disableYearSort={disableYearSort}
          noRecordFound={noRecordFound}
          sortTitle={sortTitle}
          sortPublishYear={sortPublishYear}
        />
      </main>

      <footer>
        Copyrights By Shubham Kalia (kaliashubham12@gmail.com)
      </footer>
    </div>
  )
}

export default App
