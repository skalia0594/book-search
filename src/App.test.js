import React from "react"
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import App from './App'


let container = null;
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div")
  document.body.appendChild(container)
})

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container)
  container.remove()
  container = null
})

it("should render a App", () => {
  act(() => {
    render(<App />, container)
  })
  expect(container.innerHTML).toMatchSnapshot()
})

it("Search Button click", async () => {

  act(() => {
    render(<App />, container)
  })
  // const input = document.querySelector("input")
  const button = document.querySelector("[data-testid=search]")
  expect(button.innerHTML).toBe("Search")

  const fakeData = {
    docs: [
      {
        "cover_i": 258027,
        "title": "The Lord of the Rings",
        "author_name": [
            "J. R. R. Tolkien"
        ],
        "first_publish_year": 1954,
        "key": "OL27448W",
      },
    ]
  }
  jest.spyOn(global, "fetch").mockImplementation(() =>
    Promise.resolve({
      json: () => Promise.resolve(fakeData)
    })
  )

  await act(async () => {
    button.dispatchEvent(new MouseEvent("click"))
  });
  expect(container.querySelector('input').value).toBe('')
  expect(container.querySelector('span').textContent).toBe('No Record Found')
  
})