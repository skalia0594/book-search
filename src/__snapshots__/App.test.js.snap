// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a App 1`] = `"<div class=\\"App\\"><header><div><form><label for=\\"searchBook\\">Enter Book Title: </label><input type=\\"text\\" id=\\"searchBook\\" name=\\"searchBook\\" placeholder=\\"Search for a book\\" value=\\"\\"><button type=\\"submit\\" data-testid=\\"search\\" class=\\"button\\" data-message=\\"This is search button\\">Search</button></form></div></header><main><div><div></div><ul></ul></div></main><footer>Copyrights By Shubham Kalia (kaliashubham12@gmail.com)</footer></div>"`;
